### ISSUE TITLE
[pnpm v7] crashes "add --global" with "ERROR  Unable to find the global bin directory"

### ISSUE THREAD

**Issue by lisunovdv** [pnpm/pnpm#4658]
`add --global` does not work without additional configuration

### Code to reproduce the issue:

```shell
npm install --global pnpm
pnpm add --global npm-run-all
ERROR  Unable to find the global bin directory
Run "pnpm setup" to create it automatically, or set the global-bin-dir setting, or the PNPM_HOME env variable. The global bin directory should be in the PATH.
```

### Expected behavior:
`pnpm add --global <module>` works correctly without additional configurations

### Actual behavior:
It causes an error and needs a workaround on CI

### Additional information:

```shell
##[debug] Azure DevOps
##[debug] vmImage: ubuntu-20.04
##[debug]script=npm install --global pnpm
pnpm config set store-dir /home/vsts/work/1/.pnpm-store
##[debug]pnpm -v
7.0.0
##[debug]node -v
v16.14.0
```

**Comment by josefaidt** [comment]
Hello I also began seeing this issue with pnpm v7. Using this with [fnm](https://fnm.vercel.app) we can manually set `PNPM_HOME` to point to the same location as `npm bin -g`, but previously this pointed to a cache that's loaded from npm's bin directory(?).

```console
josef ~
> fnm use 16
Using Node v16.14.2

josef ~
> npm bin -g
/Users/josef/.fnm/node-versions/v16.14.2/installation/bin
(not in PATH env variable)

josef ~
> pnpm -v
6.32.11

josef ~
> pnpm bin -g
/Users/josef/Library/Caches/fnm_multishells/96870_1651591328965/bin

josef ~
> pnpm add -g pnpm
Nothing to stop. No server is running for the store at /Users/josef/.pnpm-store/v3
Packages: +1 -1
+-
Progress: resolved 27, reused 26, downloaded 1, added 1, done

/Users/josef/.fnm/node-versions/v16.14.2/installation/pnpm-global/5:
- pnpm 6.32.11
+ pnpm 7.0.0

josef ~
> pnpm bin -g
 ERROR  Unable to find the global bin directory
Run "pnpm setup" to create it automatically, or set the global-bin-dir setting, or the PNPM_HOME env variable. The global bin directory should be in the PATH.
```

**Comment by zkochan** [comment]
This works as expected.

If you use [the standalone script](https://pnpm.io/installation#using-a-standalone-script) to install pnpm, then it will add PNPM_HOME dir to your PATH. Otherwise, you may run `pnpm setup`, as noted in the error message.

We had to remove "autodetection" of global bin location because it was causing more trouble than good.

**Comment by lisunovdv** [comment]
Hi @zkochan Maybe it needs to update the installation guide?

**Comment by caveman-dick** [comment]
Unfortunately, this has caused me a big headache with our Azure DevOps CI builds. Trying to fix using `pnpm setup`  doesn't work as it only updates the bashrc file with the new settings and doesn't update the currently running shell. I've tried manually reloading the bashrc file but that doesn't seem to work inside the script. The only way I could get this to work was to dig through the code to work out what the `pnpm setup` command was doing and manually set the PNPM_HOME and PATH variables in the script block:

```
- script: |
      export PNPM_HOME="/home/vsts/.local/share/pnpm"
      export PATH="$PNPM_HOME:$PATH"
      pnpm add -g aws-cdk
    displayName: 'Install AWS CDK'
```

This obviously is pretty painful and will need to be done for every script that is doing a global install as part of the pipeline. 

@zkochan you mentioned above that the automatic detection of the global store needed to be removed, what problems was it causing before? Is there a better workaround that I could use?

**Comment by zkochan** [comment]
>  I suddenly have issues with ts-node(npm works fine):

This is not related to this issue. This is about missing or incorrect peer dependencies.  You need to install `@types/node` and `typescript` globally because they are needed by ts-node.

**Comment by zkochan** [comment]
@caveman-dick why do you need global packages?

> @zkochan you mentioned above that the automatic detection of the global store needed to be removed, what problems was it causing before? Is there a better workaround that I could use?

There were many problems. Dozens of issues with dozens of messages. Some of the issues were that pnpm did not have write access to the global directory. Some issues were that pnpm was writing to a directory that the users didn't want to write to. Some issues were about pnpm overriding bins written by other package managers.

In general, I would recommend not using global packages in CI. Your dependencies should be declared as dev dependencies of the project.

**Comment by caveman-dick** [comment]
@zkochan we use the `aws-cdk` which just installs a cmd line tool. Locally this works fine without any global install, it picks it up from the node_modules/.bin folder, but on the CI server for some reason, this doesn't work. The only other way to get it to work is to use `npx aws-cdk`: 

![image](https://user-images.githubusercontent.com/516600/166902647-ffcc72da-ca93-4313-84a6-fa3d388578c7.png)

unfortunately `pnpx aws-cdk` (or `pnpm dlx aws-cdk`) doesn't work:

![image](https://user-images.githubusercontent.com/516600/166902750-4acefc85-c549-4991-840f-d0188f6ba0d2.png)

possibly due to the mismatch that the package is called `aws-cdk` but the executable is called `cdk`? 🤷‍♂️

Ideally, we would like to use the same package manager for all tasks to save confusion and hence why we have historically done a global install of the package. It also obviously saves having to prefix every subsequent cdk command as we run several as part of the pipeline.

**Comment by aparajita** [comment]
@zkochan, but I honestly can't understand the logic behind this implementation of the global bin dir. By default, the **global** bin directory is set to a subdirectory of the current user. How can that make sense? Other users won't be able to access or write to that directory, which on servers is a real problem. Is it your intention that each user has their own copy of the binaries?

If I set the global bin dir to a shared directory, like /usr/local/share/pnpm, pnpm absolutely refuses to find it. PNPM_HOME is definitely set and in PATH.

Any suggestions?

**Comment by joeldenning** [comment]
fwiw running `pnpm setup` didn't get rid of this error for me. To fix, I created a directory `/Users/joeldenning/pnpm` then added the following to my zshrc:

```sh
export PATH="/Users/joeldenning/pnpm:$PATH"
export PNPM_HOME="/Users/joeldenning/pnpm"
```

**Comment by zkochan** [comment]
> I honestly can't understand the logic behind this implementation of the global bin dir. By default, the global bin directory is set to a subdirectory of the current user. How can that make sense? 

This is how most other tools work (volta, nvm, etc). This is how [users suggested to do it](https://github.com/pnpm/pnpm/issues/2574)

If you think it should work differently, let's discuss it in a separate discussion.

> fwiw running pnpm setup didn't get rid of this error for me.

This looks like a bug. Was there any output?

**Comment by Deebster** [comment]
I also find that `pnpm setup` is not fixing the issue (on Win10).

(C:\Users\micro\AppData\Local\pnpm exists and is accessible and C:\Users\micro\AppData\Local\pnpm\store is populated from previous installs)

```
PS C:\Users\micro> pnpm add -g pnpm
 ERROR  Unable to find the global bin directory
Run "pnpm setup" to create it automatically, or set the global-bin-dir setting, or the PNPM_HOME env variable. The global bin directory should be in the PATH.
PS C:\Users\micro> pnpm setup
Currently 'PNPM_HOME' is set to 'C:\Users\micro\AppData\Local\pnpm'
PATH already contains PNPM_HOME

Setup complete. Open a new terminal to start using pnpm.
```
// new teminal
```
PowerShell 7.2.3
Copyright (c) Microsoft Corporation.

https://aka.ms/powershell
Type 'help' to get help.

PS C:\Users\micro> pnpm -v
7.0.0
PS C:\Users\micro> pnpm add -g pnpm
 ERROR  Unable to find the global bin directory
Run "pnpm setup" to create it automatically, or set the global-bin-dir setting, or the PNPM_HOME env variable. The global bin directory should be in the PATH.
PS C:\Users\micro> $env:PNPM_HOME = "C:\Users\micro\AppData\Local\pnpm"
PS C:\Users\micro> pnpm add -g pnpm
 ERROR  The configured global bin directory "C:\Users\micro\AppData\Local\pnpm" is not in PATH
PS C:\Users\micro> $env:Path += ";C:\Users\micro\AppData\Local\pnpm"
PS C:\Users\micro> pnpm add -g pnpm
Nothing to stop. No server is running for the store at C:\Users\micro\AppData\Local\pnpm\store\v3
The location of the currently running pnpm differs from the location where pnpm will be installed
 Current pnpm location: C:\Users\micro\AppData\Roaming\npm
 Target location: C:\Users\micro\AppData\Local\pnpm

Packages: +1
+
Packages are hard linked from the content-addressable store to the virtual store.
  Content-addressable store is at: C:\Users\micro\AppData\Local\pnpm\store\v3
  Virtual store is at:             .pnpm

C:\Users\micro\AppData\Local\pnpm\global\5:
+ pnpm 7.0.1

Progress: resolved 1, reused 0, downloaded 1, added 1, done
PS C:\Users\micro> cd C:\code\svelte\
PS C:\code\svelte> pnpm init svelte@next test
Wrote to C:\code\svelte\package.json

{
  "name": "svelte",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}
PS C:\code\svelte> pnpm -v
7.0.0
PS C:\code\svelte>
```

I suspect that restarting will fix this this but I can't at the moment.

**Comment by zkochan** [comment]
> fwiw running `pnpm setup` didn't get rid of this error for me. To fix, I created a directory `/Users/joeldenning/pnpm` then added the following to my zshrc:

Looks like `pnpm setup` cannot detect that the current shell is `zsh`. It check the value of `$SHELL` env variable but it says bash even on `zsh`. I think we can check if `$ZSH_VERSION` is set instead.

**Comment by zkochan** [comment]
Maybe this should improve it: https://github.com/pnpm/pnpm/pull/4701

**Comment by arjittw** [comment]
I my case, pnpm's global bin directory changed and it has created **C:\Users\arjit\AppData\Local\pnpm\global\5**
but not **node_modules** folder in it.
It needs the *node_modules* in it like this **C:\Users\arjit\AppData\Local\pnpm\global\5\node_modules**.  
So it throws error unable to find global bin directory, **Solved after installing anything globally** like this `$ pnpm add -g <anything>`.  This has created *node_modules* folder.  
But I have to manually delete v6 global bin and reinstall all packages.

The problem is pnpm changed global bin path so all modules just ignored like a fresh install.

**Comment by chioio** [comment]
I use the brew to install `pnpm v7.0.1`, and add global module occurred this error.
I solved by these steps:
1. Remove `PNPM_HOME` variable in `.zshrc`
2. Run `pnpm setup`
3. Cat `.zshrc`, find `PNPM_HOME` variable.
4. Run `source .zshrc`

**Hope this can help your!**

**Comment by zkochan** [comment]
Why were you removing PNPM_HOME from `.zshrc`? Did you have a bad path there? I guess we need to update `pnpm setup`, so that it can update PNPM_HOME, if it is already set but incorrectly

**Comment by ZYinMD** [comment]
A boring log of what happened to me, for information only.
1. `pnpm add -g pnpm` to update to v7.0.0
2. after a few days, got reminded v7.0.1 became available 
3. `pnpm add -g pnpm` to update to v7.0.1, encountered this issue
4. `pnpm setup` following instruction on screen
5. `pnpm add -g pnpm` again in a new terminal, issues disappears, now at v7.0.1

**Comment by apottere** [comment]
I'd like to continue using the "global" bin directory from `nvm` with `pnpm`, but it seems like that isn't an option anymore.

`nvm` automatically updates the shell path depending on which version of node you're using, which means in order to make pnpm write files there I think I need to set this in my nvmrc:
```
global-bin-dir=/Users/me/.nvm/current/bin
```

This is a symlink to whatever version of node nvm has active currently.
However, that causes pnpm to fail with this message:
```
 ERROR  The configured global bin directory "/Users/me/.nvm/current/bin" is not in PATH
```

... which is sort of true, but the directory that _resolves to_ is definitely in the PATH:
```
$ echo $PATH
...:/Users/me/.nvm/versions/node/v16.14.0/bin:....

$ ls -l ~/.nvm/current
... /Users/me/.nvm/current -> /Users/me/.nvm/versions/node/v16.14.0
```

So the result is that I can't install packages globally into different versions of node without making sure `global-bin-dir` is in sync before doing it (this includes pnpm itself)

**Comment by lizy-CZ** [comment]
"Open a new terminal to start using pnpm" is work for me.

**Comment by RafidMuhymin** [comment]
I am getting the same error when running `pnpm list -g`. And the trick of opening a new terminal/tab to use `pnpm` worked for me too.

**Comment by zkochan** [comment]
> ... which is sort of true, but the directory that resolves to is definitely in the PATH:

this we can fix.

**Comment by Garfill** [comment]
Open a new terminal did not work for me. After pnpm setup and open a new terminal, it still tell me Unable to find the global bin directory.

**Comment by zkochan** [comment]
What about the `$PNPM_HOME` env variable? Is it set?

**Comment by Garfill** [comment]
> What about the `$PNPM_HOME` env variable? Is it set?

Yes, the `$PNPM_HOME` was set as `export PNPM_HOME="/Users/me/Library/pnpm"` in .zshrc

**Comment by Garfill** [comment]
When debug in `.pnpm/pnpm@7.1.2/node_modules/pnpm/dist/pnpm.cjs`, there is not a `PNPM_HOME` property in `process.env` even though I can get this value in other .js file in other project.

**Comment by wishinghand** [comment]
I'm having the same issues on Windows Powershell. I tried to run `pnpm link --global` and receive:
```
 ERROR  Unable to find the global bin directory
Run "pnpm setup" to create it automatically, or set the global-bin-dir setting, or the PNPM_HOME env variable. The global bin directory should be in the PATH.
```

I then run `pnpm setup` and see that PNPM_HOME and Path are already set. I restart Powershell and still get the above error. I even tried a manual setting of `pnpm config set store-dir C:\.pnpm-store` and that doesn't seem to fix anything either.

**Comment by rulatir** [comment]
What is the latest version of pnpm that is completely free of this issue? I might have to lock my environment down to that version for now. Apparently it used be that pnpm reused some paths associated with npm, and this worked smoothly and automagically when running in different node environments managed by nave, each with its own npm, npm global directory, etc. etc. Now that pnpm insists on being told what to do, I must either spend a few days figuring out what to tell it, or go back to a problem-free version.

**Comment by zkochan** [comment]
V6

**Comment by Emily-0616** [comment]
`npm rm -g pnpm` works for me。
`pnpm add -g pnpm`安装成功后查询版本号为7.1.7，`npm rm -g pnpm`后再查询得到7.2.1。

**Comment by Emily-0616** [comment]
`npm rm -g pnpm` work for me.
`pnpm add -g pnpm` After the installation is successful, the query version number is 7.1.7, 
`npm rm -g pnpm` and then query to get 7.2.1

**Comment by aparajita** [comment]
@zkochan Here's the rundown of my situation:

- I'm on ubuntu 20.04.
- I installed npm via `sudo apt install nodejs`, which installs npm in /usr/bin, which requires me to use `sudo` to install packages, including `pnpm`.
- I do the following:

```bash
$ sudo npm i -g pnpm
added 1 package, and audited 2 packages in 574ms
$ sudo pnpm -v 
7.3.0
$ sudo pnpm ls -g
 ERROR  Unable to find the global bin directory
Run "pnpm setup" to create it automatically, or set the global-bin-dir setting, or the PNPM_HOME env variable. The global bin directory should be in the PATH.
$ sudo pnpm setup
Appended new lines to /root/.bashrc

Next configuration changes were made:
export PNPM_HOME="/root/.local/share/pnpm"
export PATH="$PNPM_HOME:$PATH"

Setup complete. Open a new terminal to start using pnpm.
# New terminal
$ sudo pnpm ls -g
 ERROR  Unable to find the global bin directory
Run "pnpm setup" to create it automatically, or set the global-bin-dir setting, or the PNPM_HOME env variable. The global bin directory should be in the PATH.
```

**Comment by BTBMan** [comment]
```bash
# 1.
pnpm setup

# 2.
source .zshrc
```

Works!

**Comment by JustinJsn** [comment]
When you have this issue,just

**Comment by JustinJsn** [comment]
1. Run `pnpm setup`
2. Open a new terminal, run `pnpm add -g pnpm`
3. Open a new terminal, run `pnpm --version`, got a new version.

**Comment by aparajita** [comment]
Never mind everyone, I installed `fnm` and worked around it by having a local install of node/npm/pnpm.

But it **is** a bug in pnpm that truly global installs (accessible by all users) that require privileged access do not work.

**Comment by Luckz** [comment]
Foolishly, I let pnpm update to 7.0.0 and nothing worked anymore (other than running existing software).
I'm missing any sort of migration for old installs. Meaning moving (!) the old store to the new location, while making sure previously pnpm-installed software still works.
My PATH had
`C:\Users\<me>\AppData\Roaming\npm\pnpm-global\5\node_modules\.bin` – had to remove that to be able to install things again. `pnpm setup` added a new PATH entry, but didn't remove the old one.

Considering a big part of the point behind using pnpm was saving disk space, being left with hundreds of megabytes of the old store was definitely unexpected.

There's no migration instructions, no explanation what to do to make sure your old installs use the new store, nothing to help you figure out when you can delete the old store safely without breaking tons of things.
Admittedly `pnpm install` explicitly says that it will re-create a `node_modules`, but `pnpm install .` for example just exits with a complaint.

**Comment by paxperscientiam** [comment]
I followed the basic instructions of using `npm` to uninstall `pnpm`, then reinstalling `pnpm`, then running `pnpm setup`, continuing work from a new terminal, and it seems to have worked.

One annoying issue, for me at least, is the mystery of what happened to previously installed global packages.

Is it the case that pnpm7 should be able to manage packages installed with pnpm6?

I see the store files are still at ~/.pnpm-store/v3 with the binaries at /usr/local/pnpm/bin/

If I have a question, it's this: am I supposed to nuke the store and binaries and manually reinstall my global packages after installing pnpm v7?

**Comment by Luckz** [comment]
> am I supposed to nuke the store and binaries and manually reinstall my global packages after installing pnpm v7?

The instructions are well-hidden in, say, https://github.com/pnpm/pnpm/discussions/3536#discussioncomment-2213727 :
`There's no migration script. Just reinstall the global packages.`

**Comment by dcaillibaud** [comment]
In my case pnpm was always broken by `pnpm add -g pnpm` and repaired by `npm i -g pnpm`

I'm with debian 11 and a user `foo` who handle the global dir (with `prefix=/home/foo/usr` in his `~/.npmrc`), PATH includes `/home/foo/usr/bin` for all users (shell is zsh for all).

I tried to add `global-bin-dir=/home/foo/usr/bin` in ~/.npmrc, `PNPM_HOME=/home/foo/.local/share/pnpm` in ~/.zshenv, could help `foo` but or others users can't access to this folder, so pnpm is always broken for foo or others users or both.

Specifying a custom folder `PNPM_HOME=/home/foo/pnpm` doesn't change anything, pnpm still use /home/foo/.local/share/pnpm

So I need to choose between `chmod g+rX o+rX /home/foo/.local/share` or not using pnpm to manage my globals packages (currently only `npm` and `pnpm`).

I have choosen the 2nd option because 
- I prefer to let default rights (for security but almost because a system script may reset these rights to the default ones)
- I'm not very confident in having two pnpm, one in /home/foo/usr/lib/node_modules/pnpm installed by npm and other in /home/foo/.local/share/pnpm/global/5/node_modules/pnpm both used by /home/foo/usr/bin/pnpm, depending if last update of pnpm was done by npm or pnpm.

**Comment by andy00614** [comment]
only 2 steps for me:
step1: `pnpm setup`
step2: `source ~/.zshrc` (it's very important)

then `pnpm add -g <package>`
that's ok

**Comment by dcaillibaud** [comment]
As @aparajita says, this doesn't work in a multiuser environment (with one of the users handling the global store, others using it), because default rights on the global store make it private. So you have to choose between changing default rights of the OS on `~/.local` or use npm for global packages.

- All users have `prefix=/home/foo/usr` in their `~/.npmrc` file
- `foo` can write in /home/foo/usr, especially `/home/foo/usr/bin` (in the $PATH of everybody) and `/home/foo/usr/lib/node_module`, so only him can manage global packages
- others users can read /home/foo/usr, so they can use these global packages.

It works with npm but I didn't find out how doing this with pnpm without changing default security rules on `/home/foo/.local` (where pnpm put it stuff, local packages and global ones, that's my pb).

It's not so important, managing my 2~3 global packages with npm is not really a pb for me, I was asking just in case I didn't saw the good solution…

**Comment by wangrongding** [comment]
> 

I did this and it didn't work...🥲

**Comment by onurusluca** [comment]
Having the same issue on **Windows 11 with Poweshell in Windows Terminal.**

Even after trying almost everything here, it still gave the same error.
I deleted everything about pnpm as per the instructions [here](https://pnpm.io/uninstall), but still the same.

It doesn't happen on my work pc, and I don't know why.

PNPM_HOME environment variables: **C:\Users\USERNAME\AppData\Local\pnpm**
In "Path" : **%PNPM_HOME%**

**Comment by wongshennan** [comment]
not sure if it helps for everyone

```bash
pnpm config set store-dir $PNPM_HOME
```

**Comment by wood3n** [comment]
I use windows terminal with powershell on windows10-19044, and that's my solution:

1. run `pnpm setup`
2. add `PNPM_HOME` to global environment variables, not user environment variables!!!
3. add `%PNPM_HOME%`  to `Path` 
4. restart windows terminal

![image](https://user-images.githubusercontent.com/31716713/189516929-1bda2f2c-1e6a-4727-88c7-126b4f41f6f0.png)

**Comment by 0xlsca** [comment]
Since I'm half-baked / still unexperienced with what I'm doing, the following should be taken with a big grain of salt.
However, I was able to identify an issue with the naming of the `PNPM_HOME` variable and Windows' variable expansion order.

The following mentioned variables are **user**-variables, meaning no meddling with system-wide variables.

When inspecting the contents of `PATH` via `echo %PATH%`, the output looked like `[...];%PNPM_HOME%;[...]`, which was suspicious to me.
The path was resolved correctly in the Environment Variables dialog box, but why not on the command line?

Well, it appearst that Windows resolves the variables in lexigraphical order and recursively expands them as expected - **but only after a variable has been declared, it will be expanded!**

I was able to prove my suspiction by renaming `PNPM_HOME` to `_PNPM_HOME` and adjusting `PATH` respectively. `%PNP_HOME%` vanished from the output and the correct path took its place.

**HOWEVER**: pnpm now complains about the missing `PNP_HOME` variable, which makes sense since we just renamed it.

My solution to this was adding back `PNPM_HOME` with a value of `%_PNPM_HOME%`. The variable now exists, but the path is only configured once (in `_PNP_HOME`)

**And it worked!**

***Drawback***:  In case pnpm itself or any other program updates `PNPM_HOME`, say, with the intention of changing the path to another directory, it will not work an also break pnpm.

**Comment by Nukiloco** [comment]
@0xlsca
This doesn't seem to work for me as PNPM is in npm\pnpm-global, and even though it says its installing packages globally.
Those packages just don't exist at all.

PNPM is thinking its in C:\Users\userAppData\Local\pnpm\global\5\node_modules, but when I run any global packages, it can't find the module and says its in C:\Users\user\AppData\Roaming\npm\pnpm-global\5\node_modules.

I have no clue why it's doing all of this. PNPM was working fine months ago and now it's completely just broken.
I probably have to switch to yarn or npm for a while until they fix this.

**Comment by zkochan** [comment]
@0xlsca so the issue is when someone installs pnpm with npm and then tries to manually set PNPM_HOME on Windows? What are the steps to reproduce the issue and what would be the fix?

**Comment by macluiggy** [comment]
> "Open a new terminal to start using pnpm" is work for me.

it worked hahaha, you just need to execute `pnpm setup` and then open a new terminal

**Comment by JasonGoemaat** [comment]
The directory that PNPM complains about exists.  I've run pnpm setup before and it was created.  Why exactly does it have to be in my PATH?   I've added the PNPM_HOME environment variable and added it to my path in .bashrc, but I get the error still even though it shows as the last directory in 'echo $PATH'.

**Comment by qiutian00** [comment]
> "Open a new terminal to start using pnpm" is work for me.

Thank you!

**Comment by cjosue15** [comment]
> 3\. %PNPM_HOME%

thats solved my problem on windows, thanks :)

**Comment by cjosue15** [comment]
> I use windows terminal with powershell on windows10-19044, and that's my solution:
> 
> 1. run `pnpm setup`
> 2. add `PNPM_HOME=C:\Users\[your system username]\AppData\Local\pnpm` to global `system variables`, not user variables!!!
> 3. add `%PNPM_HOME%`  to `Path` of `system variables`
> 4. restart windows terminal
> 
> ![image](https://user-images.githubusercontent.com/31716713/189516929-1bda2f2c-1e6a-4727-88c7-126b4f41f6f0.png)
> 
> 2022/09/22, update for macos-12.2.1 solutions:
> 
> 1. run `pnpm setup`, this will print `PNPM_HOME` and `PATH` in console, such as:
> 
> ```shell
> export PNPM_HOME="/Users/xxxx/Library/pnpm"
> export PATH="$PNPM_HOME:$PATH"
> ```
> 
> 2. copy these two statements and paste them to `~/.zshrc` file
> 
> ```shell
> open ~/.zshrc
> source ~/.zshrc
> ```
> 
> <img alt="image" width="541" src="https://user-images.githubusercontent.com/31716713/191649151-7d9da1b2-c133-4f2c-b41f-af30b2c87a27.png">
> 
> 3. restart a new terminal



> I use windows terminal with powershell on windows10-19044, and that's my solution:
> 
> 1. run `pnpm setup`
> 2. add `PNPM_HOME=C:\Users\[your system username]\AppData\Local\pnpm` to global `system variables`, not user variables!!!
> 3. add `%PNPM_HOME%`  to `Path` of `system variables`
> 4. restart windows terminal
> 
> ![image](https://user-images.githubusercontent.com/31716713/189516929-1bda2f2c-1e6a-4727-88c7-126b4f41f6f0.png)
> 
> 2022/09/22, update for macos-12.2.1 solutions:
> 
> 1. run `pnpm setup`, this will print `PNPM_HOME` and `PATH` in console, such as:
> 
> ```shell
> export PNPM_HOME="/Users/xxxx/Library/pnpm"
> export PATH="$PNPM_HOME:$PATH"
> ```
> 
> 2. copy these two statements and paste them to `~/.zshrc` file
> 
> ```shell
> open ~/.zshrc
> source ~/.zshrc
> ```
> 
> <img alt="image" width="541" src="https://user-images.githubusercontent.com/31716713/191649151-7d9da1b2-c133-4f2c-b41f-af30b2c87a27.png">
> 
> 3. restart a new terminal

thats solved my problem on windows, thanks :)

**Comment by cjosue15** [comment]
> I use windows terminal with powershell on windows10-19044, and that's my solution:
> 
> 1. run `pnpm setup`
> 2. add `PNPM_HOME=C:\Users\[your system username]\AppData\Local\pnpm` to global `system variables`, not user variables!!!
> 3. add `%PNPM_HOME%`  to `Path` of `system variables`
> 4. restart windows terminal
> 
> ![image](https://user-images.githubusercontent.com/31716713/189516929-1bda2f2c-1e6a-4727-88c7-126b4f41f6f0.png)
> 
> 2022/09/22, update for macos-12.2.1 solutions:
> 
> 1. run `pnpm setup`, this will print `PNPM_HOME` and `PATH` in console, such as:
> 
> ```shell
> export PNPM_HOME="/Users/xxxx/Library/pnpm"
> export PATH="$PNPM_HOME:$PATH"
> ```
> 
> 2. copy these two statements and paste them to `~/.zshrc` file
> 
> ```shell
> open ~/.zshrc
> source ~/.zshrc
> ```
> 
> <img alt="image" width="541" src="https://user-images.githubusercontent.com/31716713/191649151-7d9da1b2-c133-4f2c-b41f-af30b2c87a27.png">
> 
> 3. restart a new terminal

thats solved my problem on windows, thanks :)

**Comment by TemaSM** [comment]
> @0xlsca so the issue is when someone installs pnpm with npm and then tries to manually set PNPM_HOME on Windows? What are the steps to reproduce the issue and what would be the fix?

@zkochan Hi 👋
Confirming issue on 3 PCs. Here's how you can reproduce it (attached video of repro below):
1) use fresh Windows (just for clarity of experiment)
2) install Node LTS (via official .msi, or by using: [`nvs`](https://github.com/jasongin/nvs) / [`nvm-windows`](https://github.com/coreybutler/nvm-windows)
3) `npm i -g pnpm`
4) install any package with `-g` flag, example: `pnpm i -g npm-check-updates`
5) `ERROR  Unable to find the global bin directory ... Run "pnpm setup" to create it automatically ...`
6) `pnpm setup` // ENVs got updated
7) restart terminal
8) try to install any package with `-g` flag again
9) boom, error:
` ERROR  The configured global bin directory "C:\Users\User\AppData\Local\pnpm" is not in PATH`

node: `v18.12.1`
npm: `v8.19.2`
pnpm: `v7.14.2`

<details>
  <summary>Video demonstration (skip to 0:50):</summary>
  
https://user-images.githubusercontent.com/5501615/200105729-eac9408d-fc08-427c-824e-7df1935b2e9f.mp4

</details>

Hope it will help you 😃

**Comment by edgexie** [comment]
I have the same problem

**Comment by Kbdq** [comment]
I have same problem - "ERROR  The configured global bin directory "C:\Users\User\AppData\Local\pnpm" is not in PATH"
I have deleted and recreated PNPM_HOME environment variable (with the same content it had before) and it solved the issue for me.

**Comment by dawidmachon** [comment]
This solving problems with update.
Important, do not skip opening a new terminal.
1. Run pnpm setup
2. Open a new terminal, run pnpm add -g pnpm
3. Open a new terminal, run pnpm --version, got a new version.

**Comment by zkochan** [comment]
> PNPM_HOME was set as Expandable string (REG_EXPAND_SZ) but should be just String (REG_SZ)

This is done in v7.17.1
Let me know if it helped.

Also, there is a related issue here: https://github.com/pnpm/pnpm/issues/5696

**Comment by dfahlander** [comment]
I'm also stuck with this. I haven't messed up with any env vars or so - just installed pnpm [using npm](https://pnpm.io/installation#using-npm) some months ago. No manual changes of anything. It's been working great! But today I've got the following message while installing a dependency to a new package:

<img width="556" alt="image" src="https://user-images.githubusercontent.com/6791329/205592597-742bd246-5c9d-463d-ae56-d7c06b83025d.png">

Ok, so I follow the advice and:

<img width="567" alt="image" src="https://user-images.githubusercontent.com/6791329/205592785-a574831d-b03d-48d1-9701-7cfd2a9cca9e.png">

Don't know if the fact that I'm using `nvm` to switch between different versions of node could possibly be triggering changes in the global npm environment and affect pnpm?

What's the easiest workarond for this? Setting the PNPM_HOME dir to what? How do I find that out? Or should I rather just reinstall pnpm using the [standalone script](https://pnpm.io/installation#using-a-standalone-script) instead of [using npm](https://pnpm.io/installation#using-npm) ?

**Comment by my-random-username** [comment]
dont know if its related or not.

when i tried to add a package with -g flag "pnpm add -g" I got this error :
ERROR  The configured global bin directory "C:\Users\................\AppData\Local\pnpm" is not in PATH
running terminal with **_Run as admin_** solved the problem

**Comment by userAugustos** [comment]
> **_Run as admin_**

Since as admin u can run, looks like u need admin role to install globally with your pnpm, what may be not good, try to use the windows terminal with powershell or, run this directly:

$PNPM_HOME="C:\Users\user\AppData\Local\pnpm" | pnpm add -g <package>

to try to set your PATH

**Comment by jason-ha** [comment]
> > PNPM_HOME was set as Expandable string (REG_EXPAND_SZ) but should be just String (REG_SZ)
> 
> This is done in v7.17.1 Let me know if it helped.
> 
> Also, there is a related issue here: #5696

I have confirmed that 7.17.1 from a clean environment will get things properly configured.
However, if you run `pnpm setup` using 7.17.1+ one a Windows system that already has PNPM_HOME established, it will remain broken.

**Comment by AdvTechnoKing** [comment]
It's super funny moment

In system settings is ok:
![image](https://user-images.githubusercontent.com/17812651/216746322-c5c34ad2-464f-4cdd-b4be-0732b3f2fd50.png)
![image](https://user-images.githubusercontent.com/17812651/216746326-80f753a3-64c0-43e9-bc73-8851f94c8c0c.png)

But if you call `echo %PATH%` you will see the next:
![image](https://user-images.githubusercontent.com/17812651/216746386-b1b20c3a-28d2-4fa5-acab-518c3211489c.png)
![image](https://user-images.githubusercontent.com/17812651/216746397-87c69bf2-599e-437a-a523-f39dc23a99ac.png)

IDK why it happened

----------

The solution is simple, replace `%PNPM_HOME%` with a normal path `C:\Users\u\AppData\Local\pnpm`

**Comment by jason-ha** [comment]
> IDK why it happened
> 
> The solution is simple, replace `%PNPM_HOME%` with a normal path `C:\Users\u\AppData\Local\pnpm`

Prior to v7.17.1 setup created `PNPM_HOME` with special "expand" string type means it only has value after a certain point. `Path` is the same way. `Path` gets expanded just before `PNPM_HOME` is expanded because it is alphabetically before. Since `PNPM_HOME` doesn't have a value when `Path` is expended it is left with `%PNPM_HOME%`. If the initial letter in `PNPM_HOME` was anything less than P we'd likely have never noticed.

**Comment by AdvTechnoKing** [comment]
> less than P

That's what I thought at first, but then I saw the %USERPROFILE% variable in %PATH% and thought it wasn't that simple.

**Comment by jason-ha** [comment]
It is indeed not as a simple as alphabetical ordering, but the type of the value matters. `USERPROFILE` is not an expandable environment variable -- it has a simple string value.
```
C:\Users\Jason>reg query "HKCU\Volatile Environment" /v UserProfile

HKEY_CURRENT_USER\Volatile Environment
    UserProfile    REG_SZ    C:\Users\Jason
```
Notice `REG_SZ` is the type.
Compare to system with issue:
```
C:\Users\Jason>reg query "HKCU\Environment" /v PNPM_HOME

HKEY_CURRENT_USER\Environment
    PNPM_HOME    REG_EXPAND_SZ    C:\Users\Jason\AppData\Local\pnpm
```

You may fix your `PNPM_HOME` type by using:
```
reg add HKCU\Environment /v PNPM_HOME /t REG_SZ /d "%PNPM_HOME%" /f
```
which will keep its current value but overwrite the type.

I think this issue should be Closed now. It is fixed with 7.17.1's change. Rerunning setup won't fix things. So maybe it could be changed to patch the type.

**Comment by zkochan** [comment]
> unfortunately pnpx aws-cdk (or pnpm dlx aws-cdk) doesn't work:

@caveman-dick this issue was fixed. `pnpm dlx aws-cdk` works in the latest version of pnpm.

**Comment by zkochan** [comment]
This thread has too many unrelated issues. I am going to close it. If there are unsolved issues, create dedicated new issues.

**Comment by oshosh** [comment]
Also solved on windows 10. thx

**Comment by mockingjet** [comment]
I encountered the same issue too.

After I `pnpm setup`, still cannot upgrade pnpm in my VSCodeterminal (Powershell).
Then I opened another terminal window outside VSCode and upgraded successfully.
After that, I still needed to restart VSCode.
I guess VSCode doesn't detect Path variable changes or something

**Comment by isimmons** [comment]
Same issue here on Windows 10 with it telling me the path is not in PATH.  Updating to v8.6.8

Running the update command in a terminal outside VSCode gave me a clue what was happening. Not sure why it didn't tell me in VSCode but since I installed pnpm via `npm i -g pnpm` at some point long ago, the actual path to pnpm for me is `C:\Users\isimm\AppData\Roaming\npm`, not `C:\Users\isimm\AppData\Local\pnpm` so I had to manually fix `PNPM_HOME` to the right path in system settings.

Then reopen terminal inside or outside VSCode and it works.

I don't know why regular pnpm commands worked if the path was wrong. pnpm add worked in VSCode but pnpm list -g or pnpm i -g pnpm gave the error.

**Comment by juslin03** [comment]
@chioio thank you

**Comment by waldronmatt** [comment]
For some folks, simply opening a new terminal in vscode won't fix the issue after running `pnpm setup`. You might need to fully close out of vscode and restart which fixed the issue for me (windowss 11)

**Comment by DzmVasileusky** [comment]
If you installed it with `npm` you need to update it with `npm` as well. Since no global dirs were set up correctly
`npm i -g pnpm`
